rust/src/delta.rs

```diff
@@ -613,6 +669,192 @@ impl std::fmt::Debug for DeltaTable {
+    CommitRetriesExceeded { 
+        #[from]
+        inner: TransactionCommitAttemptError, 
+    },
+
+    #[error("RecordBatch is missing partition column in Delta schema.")]
+    MissingPartitionColumn,
+
+    #[error("Storage interaction failed: {source}")]
+    Storage { source: StorageError, },
+
+    #[error("DeltaTable interaction failed: {source}")]
+    DeltaTable { 
+        #[from]
+        source: DeltaTableError,
+    },
+
+    #[error("Action serialization failed: {source}")]
+    ActionSerializationFailed {
+        #[from]
+        source: serde_json::Error, 
+    }
+}
+
+/// Error that occurs when a single transaction commit attempt fails
+#[derive(thiserror::Error, Debug)]
+pub enum TransactionCommitAttemptError {
+    // NOTE: it would be nice to add a `num_retries` prop to this error so we can identify how frequently we hit optimistic concurrency retries and look for optimization paths
+    #[error("Version already exists: {source}")]
+    VersionExists { source: StorageError },
+
+    #[error("Commit Failed due to DeltaTable error: {source}")]
+    DeltaTable {
+        #[from]
+        source: DeltaTableError,
+    },
+
+    #[error("Commit Failed due to StorageError: {source}")]
+    Storage { source: StorageError, }
+}
+
+impl From<StorageError> for TransactionCommitAttemptError {
+    fn from(error: StorageError) -> Self {
+        match error {
+            StorageError::AlreadyExists(_) => TransactionCommitAttemptError::VersionExists { source: error },
+            _ => TransactionCommitAttemptError::Storage { source: error },
+        }
+    }
+}
+
+const DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS: u32 = 10_000_000;
+
+/// Options for customizing behavior of a `DeltaTransaction`
+pub struct DeltaTransactionOptions {
+    /// number of retry attempts allowed when committing a transaction
+    max_retry_commit_attempts: u32,
+}
+
+impl DeltaTransactionOptions {
+    /// Creates a new `DeltaTransactionOptions`
+    pub fn new(max_retry_commit_attempts: u32) -> Self {
+        Self { max_retry_commit_attempts, }
+    }
+}
+
+impl Default for DeltaTransactionOptions {
+    fn default() -> Self { 
+        Self { max_retry_commit_attempts: DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS }
+    }
+}
+
+/// Object representing a delta transaction
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+    options: DeltaTransactionOptions,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    /// Creates a new delta transaction.
+    /// Holds a mutable reference to the delta table to prevent outside mutation while a transaction commit is in progress.
+    /// Transaction behavior may be customized by passing an instance of `DeltaTransactionOptions`.
+    pub fn new(delta_table: &'a mut DeltaTable, options: Option<DeltaTransactionOptions>) -> Self {
+        DeltaTransaction { 
+            delta_table,
+            options: options.map_or_else(DeltaTransactionOptions::default, |o| o),
```

Comment:
If there is no transformation needed, we can use `unwrap_or` instead:

```suggestion
            options: options.unwrap_or(DeltaTransactionOptions::default),
```

Reply:
Nice. Thanks.